actors/stage/src/main/java/com/ea/orbit/actors/runtime/cloner/KryoCloner.java

```diff
@@ -40,75 +43,76 @@
+                    final Class<?> unmodifiableListClazz = Collections.unmodifiableList(new ArrayList()).getClass();
+                    final Class<?> unmodifiableSetClazz = Collections.unmodifiableSet(new HashSet()).getClass();
 
-            @Override
-            protected Map createCopy(Kryo kryo, Map original)
-            {
-                if (original.getClass() == unmodifiableMapClazz)
+                    @Override
+                    protected Collection createCopy(Kryo kryo, Collection original)
+                    {
+                        if (original.getClass() == unmodifiableListClazz || original.getClass() == unmodifiableCollectionClazz)
+                        {
+                            return new ArrayList();
+                        }
+                        if (original.getClass() == unmodifiableSetClazz)
+                        {
+                            return new LinkedHashSet();
+                        }
+                        return super.createCopy(kryo, original);
+                    }
+                });
+                kryo.addDefaultSerializer(Map.class, new MapSerializer()
                 {
-                    return new LinkedHashMap<>();
-                }
-                return super.createCopy(kryo, original);
+                    final Class<?> unmodifiableMapClazz = Collections.unmodifiableMap(new HashMap()).getClass();
+
+                    @Override
+                    protected Map createCopy(Kryo kryo, Map original)
+                    {
+                        if (original.getClass() == unmodifiableMapClazz)
+                        {
+                            return new LinkedHashMap<>();
+                        }
+                        return super.createCopy(kryo, original);
+                    }
+                });
+
+                kryo.addDefaultSerializer(Actor.class, new ImmutableObjectSerializer<Actor>(true, true));
+                kryo.addDefaultSerializer(UUID.class, new ImmutableObjectSerializer<UUID>(true, true));
+
+                return kryo;
             }
-        });
+        };
 
-        kryo.addDefaultSerializer(Actor.class, new ImmutableObjectSerializer<Actor>(true, true));
-        kryo.addDefaultSerializer(UUID.class, new ImmutableObjectSerializer<UUID>(true, true));
+        kryoPool = new KryoPool.Builder(factory).softReferences().build();
     }
 
     @Override
-    public <T> T clone(final T obj)
+    public <T> T clone(final T object)
     {
-        if (obj != null)
+        if (object != null)
         {
-            synchronized (kryo)
-            {
-                try
-                {
-                    return kryo.copy(obj);
-                } finally
-                {
-                    kryo.reset();
-                }
-            }
+            return kryoPool.run(kryo -> kryo.copy(object));
```

Comment:
Unsure if we should be invoking reset here explicitly, at least run takes care of returning the kyro instance back to the pool.